js/ui/searchbar.js

```diff
@@ -232,14 +244,54 @@ define([
       }
     },
 
+    searchHistory: function() {
+      var input = document.find(".searchbar").find(".search-box");
+      var history = [];
+      var currentIndex = 0;
+      var temporaryQuery = '';
+
+      return {
+        add: function(query) {
+          if (query != history[history.length - 1]) { // deduplicate
+            history.push(query);
+          }
+          currentIndex = history.length;
+          temporaryQuery = '';
+        },
+
+        showPrevious: function() {
+          if (currentIndex == history.length) {
+            temporaryQuery = input.value;
+            // skip previous search if we're already showing the same value
+            if (temporaryQuery == history[currentIndex-1]) {
+              currentIndex--;
+            }
+          }
+          if (currentIndex - 1 >= 0) {
+            input.value = history[--currentIndex];
+          }
+        },
+
+        showNext: function() {
+          if (currentIndex + 1 < history.length) {
+            input.value = history[++currentIndex];
+          } else if (currentIndex + 1 == history.length) {
+            currentIndex++;
+            input.value = temporaryQuery;
+          }
+        }
+      }
+    }(),
```

Comment:
Yeah, I was kind of wondering about that. I think it's probably just as easy to integrate the search history directly into the `Searchbar` prototype, instead of having it be its own little sub-object. That's where I expected it to be anyway.